cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,353 @@
+	}
+
+	jsonValues := ""
+	foundValues := false
+	for _, item := range configMapList.Items {
+		if item.Name == "istio-sidecar-injector" && item.Data != nil {
+			jsonValues, foundValues = item.Data["values"]
+			if foundValues {
+				break
+			}
+		}
+	}
+
+	if !foundValues {
+		l.logAndFatalf("Abort. Failed to find values in sidecar-injector config map: %v", configMapList)
+	}
+
+	yamlValues, err := yaml.JSONToYAML([]byte(jsonValues))
+	if err != nil {
+		l.logAndFatalf("jsonToYAML failed to parse values:\n%v\nError:\n%v", yamlValues, err)
+	}
+
+	return string(yamlValues)
+}
+
+// checkSupportedVersions checks if the upgrade cur -> tar is supported by the tool
+func checkSupportedVersions(cur, tar, versionsURI string) {
+	if cur == tar {
+		l.logAndFatalf("Abort. The current version %v equals to the target version %v.", cur, tar)
+	}
+
+	curPkgVer, err := pkgversion.NewVersionFromString(cur)
+	if err != nil {
+		l.logAndFatalf("Abort. Incorrect version: %v.", cur)
+	}
+
+	tarPkgVer, err := pkgversion.NewVersionFromString(tar)
+	if err != nil {
+		l.logAndFatalf("Abort. Incorrect version: %v.", tar)
+	}
+
+	if curPkgVer.Major != tarPkgVer.Major {
+		l.logAndFatalf("Abort. Major version upgrade is not supported: %v -> %v.", cur, tar)
+	}
+
+	if curPkgVer.Minor != tarPkgVer.Minor {
+		l.logAndFatalf("Abort. Minor version upgrade is not supported: %v -> %v.", cur, tar)
+	}
+
+	if curPkgVer.Patch == tarPkgVer.Patch {
+		l.logAndFatalf("Abort. The target version has been installed in the cluster.\n"+
+			"istioctl: %v\nIstio control plane: %v", cur, tar)
+	}
+
+	if curPkgVer.Patch > tarPkgVer.Patch {
+		l.logAndFatalf("Abort. A newer version has been installed in the cluster.\n"+
+			"istioctl: %v\nIstio control plane: %v", cur, tar)
+	}
+
+	tarGoVersion, err := goversion.NewVersion(tar)
+	if err != nil {
+		l.logAndFatalf("Abort. Failed to parse the target version: %v", tar)
+	}
+
+	compatibleMap := getVersionCompatibleMap(versionsURI, tarGoVersion, l)
+
+	curGoVersion, err := goversion.NewVersion(cur)
+	if err != nil {
+		l.logAndFatalf("Abort. Failed to parse the current version: %v", cur)
+	}
+
+	if !compatibleMap.SupportedIstioVersions.Check(curGoVersion) {
+		l.logAndFatalf("Abort. Upgrade is currently not supported: %v -> %v.", cur, tar)
+	}
+
+	l.logAndPrintf("Upgrade version check passed: %v -> %v.\n", cur, tar)
+}
+
+// retrieveControlPlaneVersion retrieves the version number from the Istio control plane
+func retrieveControlPlaneVersion(kubeClient manifest.ExecClient, istioNamespace string) string {
+	meshInfo, e := kubeClient.GetIstioVersions(istioNamespace)
+	if e != nil {
+		l.logAndFatalf("Failed to retrieve Istio control plane version, error: %v", e)
+	}
+
+	if meshInfo == nil {
+		l.logAndFatalf("Istio control plane not found in namespace: %v", istioNamespace)
+	}
+
+	for _, remote := range *meshInfo {
+		l.logAndPrintf("Control Plane - %s pod - version: %s", remote.Component, remote.Info.Version)
+	}
+	l.logAndPrint("")
```

Comment:
remove?

Reply:
This empty line is used to separate the next section. I'd like to keep it.